### PR TITLE
monitoring/ceph-mixin: Cleanup of variables, queries and tests (to fix showMultiCluster=True)

### PR DESCRIPTION
**Pull request by frittentheke**
Rendering the dashboards  with `showMultiCluster=True` allows for them to work with multiple clusters storing their metrics in a single Prometheus instance. This works via the (configurable) `cluster` label and that functionality already existed.

This commit simply fixes some inconsistencies in applying the label filters which I found after rendering and using the 
dashboards with a Prometheus instance holding metrics for multiple Ceph clusters.

There are also issues with the tests. I started working on them as well, but would like some feedback on how to best test with either `showMultiCluster` set to `True` or `False`. This would then ensure the support for multi cluster doesn't break on future changes and additions to the dashboards.

Fixes: https://tracker.ceph.com/issues/64321
Signed-off-by: Christian Rohmann <christian.rohmann@inovex.de>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by frittentheke** [comment]
I admit the PR got a little bigger than just "fixing" the queries, but I believe I somewhat stayed in context.
See commit msg for some of my reasoning.

**Comment by cloudbehl** [comment]
@frittentheke Thanks for PR and fixing the queries.

Just wanted to understand, rather than having a flag to enable the `cluster` variable why don't we have dashboards default have `cluster` variable enabled. So user doesn't need to build it to support multicluster.  Thoughts?

**Comment by cloudbehl** [comment]
Also can you attach the small recording that shows fixes you have done as part of the PR?

**Comment by frittentheke** [comment]
> @frittentheke Thanks for PR and fixing the queries.
> 
> Just wanted to understand, rather than having a flag to enable the `cluster` variable why don't we have dashboards default have `cluster` variable enabled. So user doesn't need to build it to support multicluster. Thoughts?

No objections on my part. We ourselves do use a single Prometheus instance (could also be Grafana Mimir / Thanos / Cortex ) holding metrics for multiple Ceph cluster and therefore make use of the templating via the `cluster` label.

**Comment by frittentheke** [comment]
> Also can you attach the small recording that shows fixes you have done as part of the PR?

@cloudbehl  you mean like a screen recording of me clicking through the various Grafana dashboards?

**Comment by cloudbehl** [comment]
> > @frittentheke Thanks for PR and fixing the queries.
> > Just wanted to understand, rather than having a flag to enable the `cluster` variable why don't we have dashboards default have `cluster` variable enabled. So user doesn't need to build it to support multicluster. Thoughts?
> 
> No objections on my part. We ourselves do use a single Prometheus instance (could also be Grafana Mimir / Thanos / Cortex ) holding metrics for multiple Ceph cluster and therefore make use of the templating via the `cluster` label.

That's my understanding as well, as if user is having multiple ceph pointing to single prometheus this would work out of the box for them  and they don't need to build grafana dashboards just fr doing it.  Even if single ceph cluster data in prometheus showing a cluster ID on top would do no harm.

So lets do and have this flag default enabled for all dashboards. I can help test this.

Also there are new dashboards added recently to grafana recently, can you help make sure those also hold the cluster variable and we are using same queries to fill values in those variables.

**Comment by cloudbehl** [comment]
> ik

I have never tried it, so just want to see how it looks on one or two dashboards if you can show. That would be great

**Comment by frittentheke** [comment]
@cloudbehl  

I seem to have fallen into a rabbit hole, just tryining to fix "a few" issues with multicluster ... 

I pushed a new revision now which also updates the recently added "RGW S3 Analystics" boards that came in via https://tracker.ceph.com/issues/64359. Unfortunately there was also https://github.com/ceph/ceph/pull/55314/ merged, which breaks the queries for <=Reef I suppose, so there so no simply backporting this anymore (if accepted and merged).

In any case, this PR has gotten was bigger than I expected. I gladly will provide a little recording of me browsing through the dashboard. But honestly this is not enough as a review. Especially the labels `instance` and `hostname` are used weirdly (I raised https://tracker.ceph.com/issues/64288 a while back). Then there is lots of label_replacing happening.
While it's nice to make the queries work on the most cluttered `instance` labeles (e.g. with port numbers), they could look much much cleaner, if it was simply expected from the source to provide them in a certain syntax (e.g. have Prometheus write clean `instance` labels. I tried to apply some cleaning and alignment, but this requires some good pair of eyes in reviewing the changes. I don't  want to break thing for some folks, but only help fixing the multicluster dashboards.


All in all I believe the mixins deserve a full refactoring a some point to ensure they are maintained

a) Upgrade / Switch to https://github.com/grafana/grafonnet (from https://github.com/grafana/grafonnet-lib which is deprecated)
b) Upgrade to the lastest Grafana panels and align all the boards even more in their style and naming and remove all of the boilerplate or explicit config with sane defaults where possible.
c) Review if certain patterns could not be moved into little helpers, just like the matchers are now. My first idea would be the label_replacement. Requiring so much post-processing of metrics and their labels and so much code in the generator code makes any change too risky.

**Comment by cloudbehl** [comment]
> > So lets do and have this flag default enabled for all dashboards. I can help test this.
> 
> @cloudbehl It might not be as easy. Even a single cluster would need to have the `cluster` label on its metrics then. With enabled `showMultiCluster` all/most queries will filter on this label to distinguish between clusters.
> 
> And most likely folks will not have Prometheus add this cluster label in their scraping config. The only thing the docs at https://docs.ceph.com/en/latest/mgr/prometheus/#honor-labels mention, is overriding the `instance` with a fixed value due to the responding mgr host changing. But there is no mention of a `cluster` variable.
> 
> I am wondering if there is any way to make this work opportunistically ... but I doubt it. There might have been a reason to have this as a configurable option after all ...
> 
> > Also there are new dashboards added recently to grafana recently, can you help make sure those also hold the cluster variable and we are using same queries to fill values in those variables.
> 
> I'll check again if all boards and queries are instrumented.
Thanks for looking into it.


This is something we will have soon in main branch via this PR((https://github.com/ceph/ceph/pull/54964) ), so all the new cluster will have the `cluster` label by default attached to it in prometheus metrics. 

We have seen lot of admin doing the same just to support the multi-cluster so its makes sense to add the cluster label by default to all the queries. Also as we have progressing with multi-cluster management and monitoring for ceph cluster. This should become standard to have the label so we don't need to rely on instance label.

**Comment by cloudbehl** [comment]
> @cloudbehl
> 
> I seem to have fallen into a rabbit hole, just tryining to fix "a few" issues with multicluster ...

What's the major issue that you are seeing with it?
> 
> I pushed a new revision now which also updates the recently added "RGW S3 Analystics" boards that came in via https://tracker.ceph.com/issues/64359. Unfortunately there was also #55314 merged, which breaks the queries for <=Reef I suppose, so there so no simply backporting this anymore (if accepted and merged).

can we have a different commit for the "RGW" realted dashboards in different PR so this could be possibly backport to fix potential issue? 

> 
> In any case, this PR has gotten was bigger than I expected. I gladly will provide a little recording of me browsing through the dashboard. But honestly this is not enough as a review. Especially the labels `instance` and `hostname` are used weirdly (I raised https://tracker.ceph.com/issues/64288 a while back). Then there is lots of label_replacing happening. While it's nice to make the queries work on the most cluttered `instance` labeles (e.g. with port numbers), they could look much much cleaner, if it was simply expected from the source to provide them in a certain syntax (e.g. have Prometheus write clean `instance` labels. I tried to apply some cleaning and alignment, but this requires some good pair of eyes in reviewing the changes. I don't want to break thing for some folks, but only help fixing the multicluster dashboards.




> 
> All in all I believe the mixins deserve a full refactoring a some point to ensure they are maintained
> 
> a) Upgrade / Switch to https://github.com/grafana/grafonnet (from https://github.com/grafana/grafonnet-lib which is deprecated)
I agree we should migrate.

 b) Upgrade to the lastest Grafana panels and align all the boards even more in their style and naming and remove all of the boilerplate or explicit config with sane defaults where possible. 
 
 Agreed this is much required, something I have been talking with the monitoring team for a while as well. 

Few improvement areas that I see:
1. All the dashboard needs to be revisited and see how we can reduce the count of dashboards. Like we have two cluster dashboard we have four rgw dashboard. I think some dashboard could be merged into 1 and create less confusion for admins.
2. Too many variables in some dashboard. Some are not even working and some are working but doesn't have proper filtering. 
3. Adding proper text helper for all the graphs.
4. All the graph/tables needs to be migrated to newer graphs/tables.

**Comment by frittentheke** [comment]
> > I seem to have fallen into a rabbit hole, just tryining to fix "a few" issues with multicluster ...
> 
> What's the major issue that you are seeing with it?


If you look at the changes, I also cleaned up (and hopefully did not break) quite a few queries that were not suitable for Prometheus holding data for e.g. other hosts than those hosting Ceph (e.g. listing all "instance" values for a Ceph dashboard template).

I still believe I did some good to all of them dashboards, even if there was a rewrite coming in, having a good base of working queries makes that process a lot easier. So I gladly invested the time.

**Comment by frittentheke** [comment]
> > I pushed a new revision now which also updates the recently added "RGW S3 Analystics" boards that came in via https://tracker.ceph.com/issues/64359. Unfortunately there was also #55314 merged, which breaks the queries for <=Reef I suppose, so there so no simply backporting this anymore (if accepted and merged).
> 
> can we have a different commit for the "RGW" realted dashboards in different PR so this could be possibly backport to fix potential issue?

You sure can, I'll look into it. 
I also renamed / aligned the target filenames to universally use the `radosgw-` prefix. Do you like that part or should I remove that altogether?

**Comment by cloudbehl** [comment]
I think renaming we can have a separate small PR after all is done just for squid and main branch

**Comment by frittentheke** [comment]
@cloudbehl I was about to split up the commit, but got a little confused. When were those metrics renamed in the mgr prometheus module? so e.g. `ceph_rgw_put_initial_lat_sum` -> `ceph_rgw_op_put_obj_lat_sum`?

I suppose the commit https://github.com/ceph/ceph/commit/be6ce973a66aa21b35760dc3b4271e682a4d596b renamed those metrics which results in those changes to only be required for master / squid branch?

**Comment by cloudbehl** [comment]
This is the PR: https://github.com/ceph/ceph/pull/54623 in which the naming were changed.

Yes the changes are in `Squid` and `main` branch as the PR was merged before the squid branch was cut.

**Comment by cloudbehl** [comment]
hey @frittentheke , I was testing your PR and wanted to understand if the `Job` variable is must have? How is job variable helpful in multi-cluster queries? 

image

**Comment by frittentheke** [comment]
> > a) Upgrade / Switch to https://github.com/grafana/grafonnet (from https://github.com/grafana/grafonnet-lib which is deprecated)
> > I agree we should migrate.
> 
> b) Upgrade to the lastest Grafana panels and align all the boards even more in their style and naming and remove all of the boilerplate or explicit config with sane defaults where possible.
> 
> Agreed this is much required, something I have been talking with the monitoring team for a while as well.
> 
> Few improvement areas that I see:
> 
>     1. All the dashboard needs to be revisited and see how we can reduce the count of dashboards. Like we have two cluster dashboard we have four rgw dashboard. I think some dashboard could be merged into 1 and create less confusion for admins.
> 
>     2. Too many variables in some dashboard. Some are not even working and some are working but doesn't have proper filtering.
> 
>     3. Adding proper text helper for all the graphs.
> 
>     4. All the graph/tables needs to be migrated to newer graphs/tables.

@cloudbehl  Is this then tracked as a bug / Trello card somewhere? I believe the cleanup we are discussing / working on with patch here is still makes sense for the old generator codebase. But the next step really should be a lift and shift of each an every panel into a newly created https://github.com/grafana/grafonnet based setup. If all or most of the queries are valid converting them over is rather simple.

**Comment by cloudbehl** [comment]
> > hey @frittentheke , I was testing your PR and wanted to understand if the `Job` variable is must have? How is job variable helpful in multi-cluster queries?
> 
> @cloudbehl
> 
> Well it was there already, I did not introduce it, but rather tried to apply it consistently. In the end it simply depends on how flexible you want the dashboards to be in regards to various Prometheus setups. So do you want to cover the case where someone is using a different `job` name to collect the metrics of a (or multiple) ceph clusters instead of applying a cluster label. But I suppose that is exactly what #54964 will ensure.
> 
> Also I am wondering how well a single job name would play with metrics collected from e.g. ceph-exporter (https://github.com/ceph/ceph/blob/b11e6e21f7271bc622c662de6f6c148d6da5c6fb/doc/cephadm/services/monitoring.rst#deploying-monitoring-without-cephadm). Those then have `ceph-exporter` as job name. So it might then be required to use e.g. `{job=~"ceph|ceph-exporter"}`.
> 
> In the end it could be best to remove the job label from the queries and templating altogether and by convention expect that a `cluster` label is present to distinguish between different clusters and to avoid any duplicate metrics when joining. This seems like the cleanest approach.
> 
> If you want I can make this cleanup / removal within this change here.

My 2 cents would be to remove the job variable from all queries.

As you rightly said we have cluster label which will help differentiate. As I understand, Multiple clusters can have same job name but multiple cluster will never have same cluster id and yes job name would conflict in graphs generating ` found duplicate series for the match group` error. So better to remove it.

Yes if the change is not too much we can do that in this same PR.

**Comment by cloudbehl** [comment]
> > > a) Upgrade / Switch to https://github.com/grafana/grafonnet (from https://github.com/grafana/grafonnet-lib which is deprecated)
> > > I agree we should migrate.
> > 
> > 
> > b) Upgrade to the lastest Grafana panels and align all the boards even more in their style and naming and remove all of the boilerplate or explicit config with sane defaults where possible.
> > Agreed this is much required, something I have been talking with the monitoring team for a while as well.
> > Few improvement areas that I see:
> > ```
> > 1. All the dashboard needs to be revisited and see how we can reduce the count of dashboards. Like we have two cluster dashboard we have four rgw dashboard. I think some dashboard could be merged into 1 and create less confusion for admins.
> > 
> > 2. Too many variables in some dashboard. Some are not even working and some are working but doesn't have proper filtering.
> > 
> > 3. Adding proper text helper for all the graphs.
> > 
> > 4. All the graph/tables needs to be migrated to newer graphs/tables.
> > ```
> 
> @cloudbehl Is this then tracked as a bug / Trello card somewhere? I believe the cleanup we are discussing / working on with patch here is still makes sense for the old generator codebase. But the next step really should be a lift and shift of each an every panel into a newly created https://github.com/grafana/grafonnet based setup. If all or most of the queries are valid converting them over is rather simple.

No I don't think its tracked somewhere but I can create tracker for the mentioned things.

Yes it would make sense for old generator as well. 

FYI: We can enable the `showMultiCluster` to `true` for the main branch once https://github.com/ceph/ceph/pull/54964 PR is merged and this PR is not planned to be migrated to Squid branch as the changes for multi cluster(https://github.com/ceph/ceph/pull/55574) is too much to be backported.

**Comment by cloudbehl** [comment]
> > No I don't think its tracked somewhere but I can create tracker for the mentioned things.
> 
> Yes @cloudbehl lease do and kindly mention the ID - I'd love to track that one.
> 
> > FYI: We can enable the `showMultiCluster` to `true` for the main branch once #54964 PR is merged and this PR is not planned to be migrated to Squid branch as the changes for multi cluster(#55574) is too much to be backported.
> 
> I now rebased this PR, removed the file renaming and dropped all of the job templating / label filtering. If this PR, once reviewed and in order, was merged to master, what are you plans on backporting any of the fixes / changes to Squid or Reef even?
> 
> I doubt that I can split the changes in commits that would cleanly merge to other branches. At least the renaming done in #54623 will cause issues. But I guess that's rather simple to fix for the backporting commit, right?

I think once we have the PR well tested, we can merge and backport it to Squid & Reef. As you mentioned we can do the required merge conflict resolutions during the back porting process itself.

@aaSharma14 @nizamial09 what's your thought on backport of this PR?

**Comment by cloudbehl** [comment]
Just did a round of testing around it.

We are removing ceph-cluster JSON(https://github.com/ceph/ceph/pull/55495/files#diff-7093d5e750937e011513fa42333853a5914a6801f8b4ec312ed56415e92e48f5) which should be restored I know this doesn't  have a libsonnet file which is why its getting removed while building but It the landing dashboard for ceph-cluster so should not be deleted.

Also I see in ceph-cluster-advanced dashboard we are not using doesn't have `cluster` templating variable. Is there any specific reason for that?

**Comment by cloudbehl** [comment]
@frittentheke Any updates on this?

**Comment by frittentheke** [comment]
@cloudbehl  I was on vacation, sorry about the laggy response.
I shall check out your review remarks shortly and push another update to the PR.

> We are removing ceph-cluster JSON(https://github.com/ceph/ceph/pull/55495/files#diff-7093d5e750937e011513fa42333853a5914a6801f8b4ec312ed56415e92e48f5) which should be restored I know this doesn't have a libsonnet file which is why its getting removed while building but It the landing dashboard for ceph-cluster so should not be deleted.

I shall restore that file with my next push then.


> Also I see in ceph-cluster-advanced dashboard we are not using doesn't have `cluster` templating variable. Is there any specific reason for that?

I doubt that. Let me check this out and fix it.

**Comment by cloudbehl** [comment]
@frittentheke Ohh, sorry to disturb you then, have a good time!

**Comment by frittentheke** [comment]
@cloudbehl I now went over the dashboards once more. PTAL.

I get more and more errors about Angular being used (which is deprecated).
The next iteration really has to be a rewrite of all of them boards using https://github.com/grafana/grafonnet.

**Comment by cloudbehl** [comment]
> @cloudbehl I now went over the dashboards once more. PTAL.

Thanks for the update. I'll test it again!
> 
> I get more and more errors about Angular being used (which is deprecated). The next iteration really has to be a rewrite of all of them boards using https://github.com/grafana/grafonnet.

I agree! It should be done!

File: monitoring/ceph-mixin/dashboards_out/ceph-cluster-advanced.json
Comment (cloudbehl) [inline monitoring/ceph-mixin/dashboards_out/ceph-cluster-advanced.json:3811]: Do we need this static uid? I did test on my existing system and all dashboard upgraded but not ceph-cluster-advance dashboard.

I am assuming this might be the thing that is message it up. Not finding any other thing.

**Review by cloudbehl — APPROVED** [review]
I tested and everything seems to work fine. Thanks for all the work!

@aaSharma14 can you check and verify it once.

**Comment by cloudbehl** [comment]
> I get more and more errors about Angular being used (which is deprecated). The next iteration really has to be a rewrite of all of them boards using https://github.com/grafana/grafonnet.

@aaSharma14 just making sure you are aware about it.

File: monitoring/ceph-mixin/dashboards_out/ceph-cluster-advanced.json
Comment (frittentheke) [reply monitoring/ceph-mixin/dashboards_out/ceph-cluster-advanced.json:3811]: All other boards already had random static IDs so I added one for this board as well to align them.
This is what Grafana has to say about UIDs: https://grafana.com/docs/grafana/latest/administration/provisioning/#reusable-dashboard-urls

I don't mind changing things around, my goal was simply to align things as much as possible.

**Review by aaSharma14 — APPROVED** [review]
Thanks @frittentheke , works fine for me!. Also regarding the Angular deprecated error on different panels that you mentioned, we have PRs to migrate two of them( table panel and pie chart panel ) to the newer versions. Here are the PR's for them respectively - https://github.com/ceph/ceph/pull/56509, https://github.com/ceph/ceph/pull/55757

File: monitoring/ceph-mixin/dashboards_out/ceph-cluster-advanced.json
Comment (cloudbehl) [reply monitoring/ceph-mixin/dashboards_out/ceph-cluster-advanced.json:3811]: Ack, we can have it, Maybe we didn't have it earlier is causing this issue. Lets continue having it.

**Comment by cloudbehl** [comment]
@frittentheke can you fix the test and we can merge this?

**Comment by frittentheke** [comment]
> @frittentheke can you fix the test and we can merge this?
> If you need help with it, let us know.

Yes @cloudbehl. I already started to work in fixing the tests.
It's just a bit more work than I expected this to be, as those tests seem to never been run (successfully) with multi cluster enabled. There is more to it than just adding the `cluster` label to the list of labels or extending the `summary` strings to have the cluster `$.MultiClusterSummary()` as the alerts also have to make sense.

But I am on it and shall push the changes soon. 


Thanks for following up and the commitment to accept this PR once done!

**Comment by cloudbehl** [comment]
> > @frittentheke can you fix the test and we can merge this?
> > If you need help with it, let us know.
> 
> Yes @cloudbehl. I already started to work in fixing the tests. It's just a bit more work than I expected this to be, as those tests seem to never been run (successfully) with multi cluster enabled. There is more to it than just adding the `cluster` label to the list of labels or extending the `summary` strings to have the cluster `$.MultiClusterSummary()` as the alerts also have to make sense.
> 
> But I am on it and shall push the changes soon.
> 
> Thanks for following up and the commitment to accept this PR once done!

Thanks for an update. I was looking into it and as mentioned there are no tests considering the multi-cluster flag and I know its a too much of tedious work.

I'm also blocking other two PR's so we don't raise any conflicts on this PR(https://github.com/ceph/ceph/pull/56509, https://github.com/ceph/ceph/pull/56575) as its a big one and pain to rebase.

**Comment by frittentheke** [comment]
> I'm also blocking other two PR's so we don't raise any conflicts on this PR(#56509, #56575) as its a big one and pain to rebase.

@cloudbehl  I have not forgotten about this one ... I shall try and push something over the weekend.
Get ready to review :see_no_evil:

**Comment by cloudbehl** [comment]
@frittentheke  Cool, Thanks!

**Comment by frittentheke** [comment]
@cloudbehl I am not quite done (let's say 90%), but I though I'd push what I got so you can start reviewing.
There also also some queries that I can't get my head around yet on how to add grouping by `cluster`. e.g. the one for the  "CephObjectMissing" alert.


In any case, PTAL - I shall update the rest as best I can tomorrow.

**Comment by cloudbehl** [comment]
Hi @frittentheke , Thanks for update. Can we separate out the effort for the alerts into different PR. Why I say so, right now every cluster has a alert manager and the alerts are evaluated for single cluster with cluster label before being pushed via federation.  So the alerts are coming correctly just that it doesn't say from which cluster is coming from in description but which is not a blocker as it can be found out in labels.

With the different PR of alerts, we need to evaluate each alert and make sure its working perfectly and it will have lot of testing effort, as mentioned by you as well. But the work for adding filter for cluster in every alert will be really required when we want to remove all the alert managers in all attached clusters and only evaluate alerts in single cluster.

**Comment by frittentheke** [comment]
> Hi @frittentheke , Thanks for update. Can we separate out the effort for the alerts into different PR. Why I say so, right now every cluster has a alert manager and the alerts are evaluated for single cluster with cluster label before being pushed via federation. So the alerts are coming correctly just that it doesn't say from which cluster is coming from in description but which is not a blocker as it can be found out in labels.
> 
> With the different PR of alerts, we need to evaluate each alert and make sure its working perfectly and it will have lot of testing effort, as mentioned by you as well. But the work for adding filter for cluster in every alert will be really required when we want to remove all the alert managers in all attached clusters and only evaluate alerts in single cluster.

Good idea @cloudbehl, the PR was just getting bigger and bigger with no end in sight :see_no_evil: 

I split off the alert related changes into https://github.com/ceph/ceph/pull/56849 and pushed another revision to this PR here removing all changes towards alerts. FYI, to show there are multiple PRs I removed the "Fixes:" statement from the commit message of this PR and only put it into the one for the alerts.


PTAL.

**Comment by cloudbehl** [comment]
> Good idea @cloudbehl, the PR was just getting bigger and bigger with no end in sight 🙈

Thanks for separating it out. It just looks lot cleaner and better to review.

I think now only part left is to make sure tests are passing.

**Comment by aaSharma14** [comment]
Hi @frittentheke , There are a couple of linting failures in the make check.

1.  jsonnet-lint failure can be fixed by running this cli command in the ceph-mixin folder - `jsonnetfmt -- "dashboards/host.libsonnet" > new_file.libsonnet` and `jsonnetfmt -- "dashboards/multi-cluster.libsonnet" > new_file.libsonnet`. You can copy the content of this new file in the respective libsonnet file.

2. grafana-lint failure can be fixed by adding a new blank line at the end of `environment.py` file.

**Comment by cloudbehl** [comment]
@frittentheke Hey, can we fix the items mentioned here https://github.com/ceph/ceph/pull/55495#issuecomment-2055425930. I want to take this forward and unblock other PRs from merge.

**Comment by frittentheke** [comment]
> @frittentheke Hey, can we fix the items mentioned here [#55495 (comment)](https://github.com/ceph/ceph/pull/55495#issuecomment-2055425930). I want to take this forward and unblock other PRs from merge.

Yes @cloudbehl, I shall try an do so in the next few hours and then push another revision.

@aaSharma14  thanks a bunch for the help! Keep it coming if there are more issues you know how to fix.
This very much extends to  https://github.com/ceph/ceph/pull/56849 which still has lots of failing tests unfortunately.

**Comment by frittentheke** [comment]
@aaSharma14 you said
> 2\. grafana-lint failure can be fixed by adding a new blank line at the end of `environment.py` file.

but there is no `environment.py`. 


@cloudbehl 
The  failing alert test:
```
tests_alerts/test_unittests.py:18: AssertionError
-------------------------------------------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------------------------------------------
  FAILED:
    expr: "(\n  rate(ceph_osd_up[5m])\n  * on(ceph_daemon) group_left(hostname) ceph_osd_metadata\n) * 60 > 1\n", time: 1m,
        exp: {ceph_daemon="osd.0", hostname="ceph", instance="ceph:9283", job="ceph"} 1.2200000000000001E+01
        got: {ceph_daemon="osd.0", hostname="ceph", instance="ceph:9283", job="ceph"} 1.21E+01

```

seems somewhat unrelated to my changes. 
@aaSharma14 last changed the expected value in this particular test 3 years ago.

File: monitoring/ceph-mixin/tests_dashboards/features/environment.py
Comment (aaSharma14) [inline monitoring/ceph-mixin/tests_dashboards/features/environment.py:136]: @frittentheke , here's the environment.py file am talking about. This needs to have a blank line at the end.

Comment (frittentheke) [reply monitoring/ceph-mixin/tests_dashboards/features/environment.py]: Uhh how embarrasing ... I simply had a typo in my `find` command ... just looked at the shell history :-(
Fixed by adding a newline now.

**Comment by aaSharma14** [comment]
jenkins test windows

**Comment by aaSharma14** [comment]
jenkins test dashboard

**Comment by aaSharma14** [comment]
jenkins test dashboard cephadm

**Comment by frittentheke** [comment]
@cloudbehl @aaSharma14 is there anything left for me to do for this PR?

**Comment by aaSharma14** [comment]
> @cloudbehl @aaSharma14 is there anything left for me to do for this PR?

@frittentheke , The PR Looks Good to me. Nothing left as such. I will try to merge it soon.

**Comment by aaSharma14** [comment]
jenkins test dashboard cephadm

**Comment by aaSharma14** [comment]
jenkins test dashboard

**Comment by aaSharma14** [comment]
Hey @frittentheke , Can you please open the backport PR's for squid and reef branches? Let me know if you need any help with that.

**Comment by frittentheke** [comment]
> Hey @frittentheke , Can you please open the backport PR's for squid and reef branches? Let me know if you need any help with that.

Gladly. Could you please let me know what I'd need to do to properly open those PRs?

**Comment by aaSharma14** [comment]
@frittentheke , There are two related issues added to this tracker - https://tracker.ceph.com/issues/64321, one tracker is for squid branch and the second one is for reef branch. Steps to open backport PR's are -

1. Checkout the squid branch
2. Do `cd src/script`
3. Run `./ceph-backport.sh --setup`
4. The script will ask you to verify or enter some details like redmine key, github username, github token etc.
5. If the setup is okay, it should return - `ceph-backport.sh: setup is OK`
6. Now run `./ceph-backport.sh <squid_tracker_number>` for eg. `./ceph-backport.sh 65838`
7. This will open the backport PR for squid
8. Similarly you can do it for reef backport as well.

**Comment by frittentheke** [comment]
> @frittentheke , There are two related issues added to this tracker - https://tracker.ceph.com/issues/64321, one tracker is for squid branch and the second one is for reef branch. Steps to open backport PR's are -
> 
>     1. Checkout the squid branch
>     2. Do `cd src/script`
>     3. Run `./ceph-backport.sh --setup`
>     4. The script will ask you to verify or enter some details like redmine key, github username, github token etc.
>     5. If the setup is okay, it should return - `ceph-backport.sh: setup is OK`

Done.

>     6. Now run `./ceph-backport.sh <squid_tracker_number>` for eg. `./ceph-backport.sh 65838`
>     7. This will open the backport PR for squid

@aaSharma14 
I suppose I should actually cherry-pick and adjust the commit to be backported, right?
Or is that done automagically?




Also it refuses due to the issues being yours:

```
ceph-backport.sh: my Redmine username is crohmann (ID 12304)
ceph-backport.sh: ERROR: https://tracker.ceph.com/issues/65838 is assigned to someone else: Aashish Sharma (ID 11319)
ceph-backport.sh: (my ID is 12304)
ceph-backport.sh: Cowardly refusing to continue
```

**Comment by aaSharma14** [comment]
> > @frittentheke , There are two related issues added to this tracker - https://tracker.ceph.com/issues/64321, one tracker is for squid branch and the second one is for reef branch. Steps to open backport PR's are -
> > ```
> > 1. Checkout the squid branch
> > 2. Do `cd src/script`
> > 3. Run `./ceph-backport.sh --setup`
> > 4. The script will ask you to verify or enter some details like redmine key, github username, github token etc.
> > 5. If the setup is okay, it should return - `ceph-backport.sh: setup is OK`
> > ```
> 
> Done.
> 
> > ```
> > 6. Now run `./ceph-backport.sh <squid_tracker_number>` for eg. `./ceph-backport.sh 65838`
> > 7. This will open the backport PR for squid
> > ```
> 
> @aaSharma14 I suppose I should actually cherry-pick and adjust the commit to be backported, right? Or is that done automagically?
> 
> Also it refuses due to the issues being yours:
> 
> ```
> ceph-backport.sh: my Redmine username is crohmann (ID 12304)
> ceph-backport.sh: ERROR: https://tracker.ceph.com/issues/65838 is assigned to someone else: Aashish Sharma (ID 11319)
> ceph-backport.sh: (my ID is 12304)
> ceph-backport.sh: Cowardly refusing to continue
> ```

@frittentheke , The cherry-pick is done automatically with this script. However if there are any conflicts, you can just resolve the conflicts..Do a `git add` and then re-run the script and that should be it. Also i have changed the assignee to you, You can try again. Thanks

**Comment by aaSharma14** [comment]
Thank you for the backport @frittentheke , Can you please open the reef backport for this as well?

**Pull request merged**
